File: frontend/src/services/api/batch-transfer.ts
// API client for F4.2 generic batch privacy transfers.
// Aligned with backend handlers/batch_transfer.rs + services/batch_transfer_service.rs.

import api from './axios';
import type {
  BatchTransferRun,
  BatchTransferRunSummary,
  CreateBatchTransferRunRequest,
  CreateBatchTransferRunResponse,
  ExecuteBatchTransferOutcome,
} from '../../types/batch-transfer';
import { generateIdempotencyKey } from './approval';

export const batchTransferService = {
  async listRuns(limit = 50, offset = 0): Promise<BatchTransferRun[]> {
    return api.get('/batch-transfers', { params: { limit, offset } });
  },

  async getRun(id: number): Promise<BatchTransferRunSummary> {
    return api.get(`/batch-transfers/${id}`);
  },

  /**
   * Validates every imported row server-side; a non-empty
   * `validation_errors` means nothing was created (run_id = 0) and the
   * whole CSV can be fixed in one pass.
   */
  async createRun(req: CreateBatchTransferRunRequest): Promise<CreateBatchTransferRunResponse> {
    return api.post('/batch-transfers', req, {
      headers: { 'Idempotency-Key': generateIdempotencyKey() },
    });
  },

  /**
   * Starts the run or arms the staggered schedule. Same F2.1 semantics as
   * migrations: one approval covers the whole window, cancel is the only
   * way to stop remaining items.
   */
  async executeRun(id: number): Promise<ExecuteBatchTransferOutcome> {
    return api.post(`/batch-transfers/${id}/execute`, null, {
      headers: { 'Idempotency-Key': generateIdempotencyKey() },
    });
  },

  async approveRun(id: number): Promise<BatchTransferRun> {
    return api.post(`/batch-transfers/${id}/approve`, null, {
      headers: { 'Idempotency-Key': generateIdempotencyKey() },
    });
  },

  async rejectRun(id: number, reason: string): Promise<BatchTransferRun> {
    return api.post(`/batch-transfers/${id}/reject`, { reason });
  },

  async cancelRun(id: number): Promise<BatchTransferRun> {
    return api.post(`/batch-transfers/${id}/cancel`);
  },

  async retryItem(runId: number, itemId: number): Promise<BatchTransferRunSummary> {
    return api.post(`/batch-transfers/${runId}/items/${itemId}/retry`, null, {
      headers: { 'Idempotency-Key': generateIdempotencyKey() },
    });
  },
};

File: backend/src/api/handlers/batch_transfer.rs
//! F4.2 — batch privacy transfer endpoints (PRD-F4 §5). Mirrors the
//! migration endpoints one-for-one; admin-only for the same reason —
//! these move treasury funds.

use std::sync::Arc;

use actix_web::{web, HttpResponse};
use serde::Deserialize;

use crate::api::middleware::AuthenticatedUser;
use crate::db::models_f4::{CreateBatchTransferRunRequest, RejectBatchTransferRunRequest};
use crate::error::{AppError, AppResult};
use crate::services::BatchTransferService;

fn require_admin(user: &AuthenticatedUser) -> AppResult<()> {
    if user.role != "admin" {
        return Err(AppError::Forbidden(
            "Only admin can manage batch transfer runs".to_string(),
        ));
    }
    Ok(())
}

/// POST /batch-transfers — validate the imported rows and create a plan
pub async fn create_batch_transfer_run(
    service: web::Data<Arc<BatchTransferService>>,
    user: AuthenticatedUser,
    request: web::Json<CreateBatchTransferRunRequest>,
) -> AppResult<HttpResponse> {
    require_admin(&user)?;
    let response = service.create_run(request.into_inner(), user.user_id).await?;
    Ok(HttpResponse::Ok().json(response))
}

#[derive(Debug, Deserialize)]
pub struct ListQuery {
    pub limit: Option<i32>,
    pub offset: Option<i32>,
}

/// GET /batch-transfers — list runs
pub async fn list_batch_transfer_runs(
    service: web::Data<Arc<BatchTransferService>>,
    user: AuthenticatedUser,
    query: web::Query<ListQuery>,
) -> AppResult<HttpResponse> {
    require_admin(&user)?;
    let runs = service
        .list_runs(query.limit.unwrap_or(50), query.offset.unwrap_or(0))
        .await?;
    Ok(HttpResponse::Ok().json(runs))
}

/// GET /batch-transfers/{id} — run + items progress
pub async fn get_batch_transfer_run(
    service: web::Data<Arc<BatchTransferService>>,
    user: AuthenticatedUser,
    path: web::Path<i32>,
) -> AppResult<HttpResponse> {
    require_admin(&user)?;
    let summary = service.get_run_summary(path.into_inner()).await?;
    Ok(HttpResponse::Ok().json(summary))
}

/// POST /batch-transfers/{id}/execute — start now or arm the schedule.
/// May pivot to awaiting_approval per F2.1 policy.
pub async fn execute_batch_transfer_run(
    service: web::Data<Arc<BatchTransferService>>,
    user: AuthenticatedUser,
    path: web::Path<i32>,
) -> AppResult<HttpResponse> {
    require_admin(&user)?;
    let outcome = service.execute_run(path.into_inner(), user.user_id).await?;
    Ok(HttpResponse::Ok().json(outcome))
}

/// POST /batch-transfers/{id}/approve — checker approval (maker ≠ checker
/// enforced at the SQL layer). The approval covers the whole window.
pub async fn approve_batch_transfer_run(
    service: web::Data<Arc<BatchTransferService>>,
    user: AuthenticatedUser,
    path: web::Path<i32>,
) -> AppResult<HttpResponse> {
    require_admin(&user)?;
    let run = service.approve_run(path.into_inner(), user.user_id).await?;
    Ok(HttpResponse::Ok().json(run))
}

/// POST /batch-transfers/{id}/reject — checker rejection with a reason
pub async fn reject_batch_transfer_run(
    service: web::Data<Arc<BatchTransferService>>,
    user: AuthenticatedUser,
    path: web::Path<i32>,
    request: web::Json<RejectBatchTransferRunRequest>,
) -> AppResult<HttpResponse> {
    require_admin(&user)?;
    let run = service
        .reject_run(path.into_inner(), user.user_id, &request.reason)
        .await?;
    Ok(HttpResponse::Ok().json(run))
}

/// POST /batch-transfers/{id}/cancel — the only way to stop remaining items
pub async fn cancel_batch_transfer_run(
    service: web::Data<Arc<BatchTransferService>>,
    user: AuthenticatedUser,
    path: web::Path<i32>,
) -> AppResult<HttpResponse> {
    require_admin(&user)?;
    let run = service.cancel_run(path.into_inner()).await?;
    Ok(HttpResponse::Ok().json(run))
}

/// POST /batch-transfers/{run_id}/items/{item_id}/retry — replay one failed item
pub async fn retry_batch_transfer_item(
    service: web::Data<Arc<BatchTransferService>>,
    user: AuthenticatedUser,
    path: web::Path<(i32, i32)>,
) -> AppResult<HttpResponse> {
    require_admin(&user)?;
    let (run_id, item_id) = path.into_inner();
    service.retry_item(run_id, item_id).await?;
    let summary = service.get_run_summary(run_id).await?;
    Ok(HttpResponse::Ok().json(summary))
}
